cirq-core/cirq/transformers/transformer_primitives.py

```diff
@@ -275,8 +286,16 @@ def apply_merge_func(op1: ops.Operation, op2: ops.Operation) -> Optional[ops.Ope
                     ).with_tags(*op.tags, _circuit_op_tag)
                 )
                 continue
+
+            op_ck = set(op.classical_controls)
+
+            def get_idx(op_qs: Set['cirq.Qid']):
+                op = ops.identity_each(*op_qs).with_classical_controls(*op_ck)
+                idx = ret_circuit.earliest_available_moment(op) - 1
+                return idx if (idx >= 0 and ret_circuit[idx].operates_on(op_qs)) else None
```

Comment:
Having to make a dummy operation suggests to me that `Circuit.prev_moment_operation_on` and related methods should be augmented to support classical controls instead.

Reply:
+1 I think `prev_moment_operating_on` needs to accept optional `measurement_keys` and `control_keys` parameters and use those.

Alternately it may be worth it here not to use those functions since going back moment-by-moment is n^2 and hydrating immutable moments one op at a time is slow. I redesigned circuit construction in #5332 this way and got a huge speedup there, so maybe something similar could be done here.

Reply:
I took Dax's advice and rewrote the implementation avoiding circuit's built-in methods for iteration and explicitly maintaining data structures for a more optimized implementation. 

The total running time of `pytest cirq-core/cirq/transformers/transformer_primitives_test.py` reduced from 16seconds to 5 seconds on my local machine. 

Also added tests for not allowing measurements to move past conflicting CCOs.